### ISSUE TITLE
why torch.nn.linear op is much faster than torch.matmul even when they have same calculating complexity?

### ISSUE THREAD

**Issue by HudashiNeo** [pytorch/pytorch#113934]
### 🐛 Describe the bug

why linear op is so much faster than multiplication ?

if I use multiplication(@ op), how could i speed up so that i could get almost same performance like torch.nn.linear?

here is my testcase:

import torch
import time

batch, dim = 100, 768
matrix = torch.randn(batch, dim, dim)
vector = torch.randn(batch, dim)
linear = torch.nn.Linear(dim, dim)

times = 100

then = time.time()
for i in range(times):
    out0 = (matrix @ vector.unsqueeze(-1)).squeeze(-1)
print(time.time() - then)
#
then = time.time()
for i in range(times):
    out1 = linear(vector)
print(time.time() - then)
#
#
output:
time of multiplying:  8.834540128707886
time of linear op:  0.44869112968444824
#

it turns out that linear op is almost 20times faster than multiplying.

but, let's see their calculating complexity: I think both of the are (batch * dim * dim) right? we got a output vector of batch*dim, each value of which takes dim times multiplication op. 

but why linear op is so much faster than multiplication ?

if I use multiplication(@ op), how could i speed up so that i could get almost same performance like torch.nn.linear?

### Versions

PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.24.1
Libc version: N/A

Python version: 3.10.5 (v3.10.5:f377153967, Jun  6 2022, 12:36:10) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-12.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] mypy==1.6.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.3
[pip3] pytorch-crf==0.7.2
[pip3] torch==2.1.0
[pip3] torchvision==0.16.0
[conda] No relevant packages

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by HudashiNeo** [comment]
could anybody help me with this question ?

**Comment by albanD** [comment]
hey!

A couple things:
- The weight matrix doesn't have a batch dimension. So your `matrix` here shouldn't either. That makes the manual call do a bunch of extra work.
- The actual computation in linear is `out02 = torch.matmul(vector, matrix.t())` which has different in-memory layout and thus slightly different runtime behavior.

Once doing these changes, the matmul matches the linear call as expected, so closing this.
Feel free to ask follow up here or on the forum!

**Comment by HudashiNeo** [comment]
I see, sounds like that Linear op has been optimized in its "in-memory layout", that is why it runs much more fast when "batch" dim is greater. I checked that when batch <= 5 or so, @ op could be a little bit faster than linear, but when batch >= 6, time cost of @ will grow up rapidly.

**Comment by HudashiNeo** [comment]
@albanD Could there be any effective way for me to speed up @(or torch.matmul) when I have a "batch" problem? I thought it for a long time, but it is  impossible for me to use Linear to replace batch anyway, because I have to do a give each matmul op a different vector and matrix in a batch in my model. but when I do so, the time cost is really a pain point

**Comment by HudashiNeo** [comment]
> Your problem is expected to be slower because the "matrix" for linear is the same for all element in the batch while in your case, there is one for each batch sample. Since mm ops are very often memory bound (the limiting factor is the memory bandwidth not the compute cores), then being in a situation where you have an order of magnitude more data to load is going to slow it down.

@albanD so in my case(matrix for each sample in a batch), the memory bandwidth issue happens both in CPU & GPU environment ?

**Comment by albanD** [comment]
The memory bandwidth difference is the same in both cases but the ration memory bandwidth / compute is very different between CPU and GPU: where GPU are fully memory bound in this kind of cases. So the impact will be a lot more significant on GPU compared to CPU.

**Comment by HudashiNeo** [comment]
> The memory bandwidth difference is the same in both cases but the ration memory bandwidth / compute is very different between CPU and GPU: where GPU are fully memory bound in this kind of cases. So the impact will be a lot more significant on GPU compared to CPU.

@albanD I see, thank you for your patience and kindly responses for my question all the way!
